tests/test_run.py

```diff
@@ -89,3 +92,76 @@ def test_main(self):
 
         mock_args.action = 'action-not-found'
         self.assertRaises(exceptions.ActionNotFound, run.main, mock_args)
+
+    def test_invalid_dbdir(self):
+        """Check invalid dbdir raises an exception."""
+        mock_args = mock.Mock()
+        mock_args.action = 'generate'
+        mock_args.tree = 'rhel7'
+        mock_args.kernel = 'kernel'
+        mock_args.arch = 'arch'
+        mock_args.db = '/tmp/igotgoosebutnoshoes'
+        mock_args.output = None
+        mock_args.pw_cookie = None
+        mock_args.description = 'description'
+        mock_args.mboxes = []
+
+        with self.assertRaises(Exception):
+            run.main(mock_args)
+
+    def test_print_testcases(self):
+        """Check invalid dbdir raises an exception."""
+        mock_args = mock.Mock()
+        mock_args.action = 'print-test-cases'
+        mock_args.tree = 'rhel7'
+        mock_args.kernel = 'kernel'
+        mock_args.arch = 'arch'
+        mock_args.db = os.path.join(os.path.dirname(__file__), 'assets')
+        mock_args.output = None
+        mock_args.pw_cookie = None
+        mock_args.description = 'description'
+        mock_args.mboxes = []
+        mock_args.patches = []
+
+        with mock.patch('sys.stdout') as mock_stdout:
+            run.main(mock_args)
+            self._assert_standard_testcases(mock_stdout)
+
+    def test_invalid_tree(self):
+        """Check invalid tree raises an exception."""
+
+        mock_args = mock.Mock()
+        mock_args.action = 'generate'
+        mock_args.tree = 'rhel0'
+        mock_args.kernel = 'kernel'
+        mock_args.arch = 'arch'
+        mock_args.db = os.path.join(os.path.dirname(__file__), 'assets')
+        mock_args.output = None
+        mock_args.pw_cookie = None
+        mock_args.description = 'description'
+        mock_args.mboxes = []
+
+        with self.assertRaises(Exception):
+            run.main(mock_args)
+
+    def _assert_standard_testcases(self, mock_stdout):
+        self.assertEqual(
+            mock_stdout.write.call_args_list[0],
+            mock.call('default/ltplite'),
+        )
+        self.assertEqual(
+            mock_stdout.write.call_args_list[2],
+            mock.call('fs/ext4'),
+        )
+        self.assertEqual(
+            mock_stdout.write.call_args_list[4],
+            mock.call('fs/xfs'),
+        )
+
+    def test_get_test_cases_print(self):
+        """Check method prints output as it should."""
+        database = data.Base(self.dbdir)
+
+        with mock.patch('sys.stdout') as mock_stdout:
+            run.print_test_cases([], database)
+            self._assert_standard_testcases(mock_stdout)
```

Comment:
I wonder if you could just substitute sys.stdout with a StringIO and then compare the whole written string instead of messing with write calls, or do something along those lines?